x-pack/plugins/security_solution/public/detection_engine/rule_monitoring/components/basic/filters/multiselect_filter/index.tsx

```diff
@@ -0,0 +1,121 @@
+  const MultiselectFilterComponent: React.FC<MultiselectFilterProps<T>> = (props) => {
+    const {
+      dataTestSubj,
+      title,
+      items,
+      itemsSelectedByDefault,
+      onSelectionChange,
+      renderItem,
+      renderLabel,
+    } = initializeProps(props);
+
+    const [selectedItems, setSelectedItems] = useState<T[]>(itemsSelectedByDefault);
+    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+
+    const handleItemClick = useCallback(
+      (item: T) => {
+        setSelectedItems(
+          selectedItems.includes(item)
+            ? selectedItems.filter((i) => i !== item)
+            : [...selectedItems, item]
+        );
+      },
+      [selectedItems]
+    );
+
+    const filterItemElements = useMemo(() => {
+      return items.map((item, index) => {
+        const itemLabel = renderLabel(item);
+        const itemElement = renderItem(item);
+        return (
+          <EuiFilterSelectItem
+            data-test-subj={`${dataTestSubj}-item`}
+            title={itemLabel}
+            key={`${index}-${itemLabel}`}
+            checked={selectedItems.includes(item) ? 'on' : undefined}
+            onClick={() => handleItemClick(item)}
+          >
+            {itemElement}
+          </EuiFilterSelectItem>
+        );
+      });
+    }, [dataTestSubj, items, selectedItems, renderItem, renderLabel, handleItemClick]);
+
+    useEffect(() => {
+      onSelectionChange(selectedItems);
+    }, [onSelectionChange, selectedItems]);
+
+    return (
+      <EuiFilterGroup data-test-subj={dataTestSubj}>
+        <EuiPopover
+          data-test-subj={`${dataTestSubj}-popover`}
+          button={
+            <EuiFilterButton
+              data-test-subj={`${dataTestSubj}-popoverButton`}
+              iconType="arrowDown"
+              grow={false}
+              numFilters={items.length}
+              numActiveFilters={selectedItems.length}
+              hasActiveFilters={selectedItems.length > 0}
+              isSelected={isPopoverOpen}
+              onClick={() => setIsPopoverOpen(!isPopoverOpen)}
+            >
+              {title}
+            </EuiFilterButton>
+          }
+          isOpen={isPopoverOpen}
+          closePopover={() => setIsPopoverOpen(false)}
+          panelPaddingSize="none"
+          repositionOnScroll
+        >
+          {filterItemElements}
+        </EuiPopover>
+      </EuiFilterGroup>
+    );
+  };
+
+  const MultiselectFilter = React.memo(MultiselectFilterComponent);
+  MultiselectFilter.displayName = 'MultiselectFilter';
+
+  return MultiselectFilter;
+};
+
+const initializeProps = <T extends unknown>(
+  props: MultiselectFilterProps<T>
+): Required<MultiselectFilterProps<T>> => {
+  const onSelectionChange: (selectedItems: T[]) => void = props.onSelectionChange ?? noop;
+  const renderLabel: (item: T) => string = props.renderLabel ?? String;
+  const renderItem: (item: T) => React.ReactChild = props.renderItem ?? renderLabel;
```

Comment:
Discussed with @xcrzx. Not necessary in this case, because:
- these are values, not arrow functions created in place
- the component itself is wrapped with React.memo